### PR TITLE
Build copy model version API for SQL and UC model registry stores

### PR DESCRIPTION
**Pull request by jerrylian-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/jerrylian-db/mlflow/pull/10078?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10078/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10078
```

</p>
</details>

### Related Issues/PRs

https://github.com/mlflow/mlflow/pull/9946

### What changes are proposed in this pull request?

This PR implements the `copy_model_version` API for the SQL, workspace model registry, and unity catalog model registry stores.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

**SQL store**
Confirming that the copy model version works

Screenshot 2023-10-24 at 11 48 03 AM

Confirming that copy model version works in the same registered model
Screenshot 2023-10-24 at 11 49 05 AM

Then I tested the DB migration by first logging a model version in SQLite using the master branch. Then running the db migration on the SQLite file. And then finally, trying to load the model version. See that it worked below:

Screenshot 2023-10-24 at 11 54 44 AM

**UC model registry**
Successfully copied a UC model version
Screenshot 2023-10-24 at 11 01 47 AM

To test promoting feature store packaged models, I utilized the feature store team's smoke test. I verified that for both normal and on demand feature store models, the cloned version of the model passes the same tests (batch scoring of loaded model) as the source model.

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/resources/db/latest_schema.sql [hunk 1]
```diff
@@ -88,9 +88,9 @@ CREATE TABLE model_versions (
 
 
 CREATE TABLE registered_model_aliases (
```
Comment (jerrylian-db) [inline tests/resources/db/latest_schema.sql:90]: @BenWilson2 I'm not sure why the `storage_location` column was not added in this file when I ran `./tests/db/update_schemas.sh`. Can you help me understand?

File: tests/resources/db/latest_schema.sql [hunk 2]
```diff
@@ -166,7 +166,7 @@ CREATE TABLE model_version_tags (
 
 CREATE TABLE params (
 	key VARCHAR(250) NOT NULL,
-	value VARCHAR(8000) NOT NULL,
+	value VARCHAR(500) NOT NULL,
```
Comment (jerrylian-db) [inline tests/resources/db/latest_schema.sql:169]: This seems concerning. Don't know why this change was made by `./tests/db/update_schemas.sh`.

Comment (jerrylian-db) [reply tests/resources/db/latest_schema.sql]: Alright, I manually overrode those changes

File: tests/resources/db/latest_schema.sql (hunk 1 above)
Comment (jerrylian-db) [reply tests/resources/db/latest_schema.sql:91]: Alright, I manually added it later.

**Comment by jerrylian-db** [comment]
@mlflow-automation autoformat

File: tests/resources/db/latest_schema.sql (hunk 1 above)
Comment (jerrylian-db) [reply tests/resources/db/latest_schema.sql:90]: So `update_schemas.sh` calls `python tests/store/dump_schema.py`, which has the following output.

```
+ python tests/store/dump_schema.py tests/resources/db/latest_schema.sql
/Users/jerry.liang/mlflow/.venvs/mlflow-dev/lib/python3.8/site-packages/pydantic/_internal/_fields.py:128: UserWarning: Field "model_server_url" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/Users/jerry.liang/mlflow/.venvs/mlflow-dev/lib/python3.8/site-packages/pydantic/_internal/_config.py:317: UserWarning: Valid config keys have changed in V2:
* 'schema_extra' has been renamed to 'json_schema_extra'
  warnings.warn(message, UserWarning)
2023/10/23 15:27:45 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2023/10/23 15:27:45 INFO mlflow.store.db.utils: Updating database tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 451aebb31d03, add metric step
INFO  [alembic.runtime.migration] Running upgrade 451aebb31d03 -> 90e64c465722, migrate user column to tags
INFO  [alembic.runtime.migration] Running upgrade 90e64c465722 -> 181f10493468, allow nulls for metric values
INFO  [alembic.runtime.migration] Running upgrade 181f10493468 -> df50e92ffc5e, Add Experiment Tags Table
INFO  [alembic.runtime.migration] Running upgrade df50e92ffc5e -> 7ac759974ad8, Update run tags with larger limit
INFO  [alembic.runtime.migration] Running upgrade 7ac759974ad8 -> 89d4b8295536, create latest metrics table
INFO  [89d4b8295536_create_latest_metrics_table_py] Migration complete!
INFO  [alembic.runtime.migration] Running upgrade 89d4b8295536 -> 2b4d017a5e9b, add model registry tables to db
INFO  [2b4d017a5e9b_add_model_registry_tables_to_db_py] Adding registered_models and model_versions tables to database.
INFO  [2b4d017a5e9b_add_model_registry_tables_to_db_py] Migration complete!
INFO  [alembic.runtime.migration] Running upgrade 2b4d017a5e9b -> cfd24bdc0731, Update run status constraint with killed
INFO  [alembic.runtime.migration] Running upgrade cfd24bdc0731 -> 0a8213491aaa, drop_duplicate_killed_constraint
INFO  [alembic.runtime.migration] Running upgrade 0a8213491aaa -> 728d730b5ebd, add registered model tags table
INFO  [alembic.runtime.migration] Running upgrade 728d730b5ebd -> 27a6a02d2cf1, add model version tags table
INFO  [alembic.runtime.migration] Running upgrade 27a6a02d2cf1 -> 84291f40a231, add run_link to model_version
INFO  [alembic.runtime.migration] Running upgrade 84291f40a231 -> a8c4a736bde6, allow nulls for run_id
INFO  [alembic.runtime.migration] Running upgrade a8c4a736bde6 -> 39d1c3be5f05, add_is_nan_constraint_for_metrics_tables_if_necessary
INFO  [alembic.runtime.migration] Running upgrade 39d1c3be5f05 -> c48cb773bb87, reset_default_value_for_is_nan_in_metrics_table_for_mysql
INFO  [alembic.runtime.migration] Running upgrade c48cb773bb87 -> bd07f7e963c5, create index on run_uuid
INFO  [alembic.runtime.migration] Running upgrade bd07f7e963c5 -> 0c779009ac13, add deleted_time field to runs table
INFO  [alembic.runtime.migration] Running upgrade 0c779009ac13 -> cc1f77228345, change param value length to 500
INFO  [alembic.runtime.migration] Running upgrade cc1f77228345 -> 97727af70f4d, Add creation_time and last_update_time to experiments table
INFO  [alembic.runtime.migration] Running upgrade 97727af70f4d -> 3500859a5d39, Add Model Aliases table
INFO  [alembic.runtime.migration] Running upgrade 3500859a5d39 -> 7f2a7d5fae7d, add datasets inputs input_tags tables
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
```

See that the latest two migrations "add storage location field to model versions" and "increase max param val length from 500 to 8000" are not applied to "latest_schemas.sql".

Comment (jerrylian-db) [reply tests/resources/db/latest_schema.sql:91]: FYI @harupy and as we discussed, I will manually override `latest_schema.py` for now.

File: mlflow/store/_unity_catalog/registry/rest_store.py [hunk 3]
```diff
@@ -745,6 +745,19 @@ def get_model_version_by_alias(self, name, alias):
...
+        Copy a model version from one registered model to another as a new model version.
+
+        :param src_mv: A :py:class:`mlflow.entities.model_registry.ModelVersion` object representing
+                       the source model version.
+        :param dst_name: the name of the registered model to copy the model version to. If a
+                         registered model with this name does not exist, it will be created.
+        :return: Single :py:class:`mlflow.entities.model_registry.ModelVersion` object representing
+                 the cloned model version.
+        """
+        return self._copy_model_version_impl(src_mv, dst_name)
```
Comment (jerrylian-db) [inline mlflow/store/_unity_catalog/registry/rest_store.py:759]: For the UC model registry, setting the source to the model URI will not cause any issues.

File: mlflow/store/model_registry/databricks_workspace_model_registry_rest_store.py [hunk 4]
```diff
@@ -29,6 +29,19 @@ def get_model_version_by_alias(self, name, alias):
...
+        Copy a model version from one registered model to another as a new model version.
+
+        :param src_mv: A :py:class:`mlflow.entities.model_registry.ModelVersion` object representing
+                       the source model version.
+        :param dst_name: the name of the registered model to copy the model version to. If a
+                         registered model with this name does not exist, it will be created.
+        :return: Single :py:class:`mlflow.entities.model_registry.ModelVersion` object representing
+                 the cloned model version.
+        """
+        return self._copy_model_version_impl(src_mv, dst_name)
```
Comment (jerrylian-db) [inline mlflow/store/model_registry/databricks_workspace_model_registry_rest_store.py:43]: For the workspace model registry, I have a PR in progress that enables creating model versions using model version URIs.

File: tests/resources/db/latest_schema.sql (hunk 1 above)
Comment (jerrylian-db) [reply tests/resources/db/latest_schema.sql:91]: I ran `update_schemas.sh` again today, and somehow this time, it worked. 👀

**Comment by jerrylian-db** [comment]
@mlflow-automation autoformat

File: tests/db/check_migration.py [hunk 5]
```diff
@@ -109,7 +109,7 @@ def post_migration():
         for table in TABLES:
             df_actual = pd.read_sql(sa.text(f"SELECT * FROM {table}"), conn)
             df_expected = pd.read_pickle(SNAPSHOTS_DIR / f"{table}.pkl")
-            pd.testing.assert_frame_equal(df_actual, df_expected)
+            pd.testing.assert_frame_equal(df_actual[df_expected.columns], df_expected)
```
Comment (BenWilson2) [inline tests/db/check_migration.py:112]: great find on this @jerrylian-db

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for all of the patience with this one

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 6]
```diff
@@ -631,6 +635,18 @@ def next_version(sql_registered_model):
         _validate_model_name(name)
         for tag in tags or []:
             _validate_model_version_tag(tag.key, tag.value)
+        storage_location = source
+        if urllib.parse.urlparse(source).scheme == "models":
+            (src_model_name, src_model_version, _, _) = _parse_model_uri(source)
```
Comment (harupy) [inline mlflow/store/model_registry/sqlalchemy_store.py:640]: Does `_parse_model_uri` return a named tuple?

Comment (harupy) [reply mlflow/store/model_registry/sqlalchemy_store.py:640]: If so, accessing fields by name might be safer than unpacking.

Comment (jerrylian-db) [reply mlflow/store/model_registry/sqlalchemy_store.py]: Makes sense! I've changed the access method.

**Comment by jerrylian-db** [comment]
Update: I'm going to remove the workspace model registry implementation for now because the backend change needs time to bake and roll out.

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 7]
```diff
@@ -631,6 +635,20 @@ def next_version(sql_registered_model):
...
             _validate_model_version_tag(tag.key, tag.value)
+        storage_location = source
+        if urllib.parse.urlparse(source).scheme == "models":
+            parsed_model_uri = _parse_model_uri(source)
+            src_model_name = parsed_model_uri.name
+            src_model_version = parsed_model_uri.version
+            try:
+                storage_location = self.get_model_version_download_uri(
+                    src_model_name, src_model_version
+                )
```
Comment (harupy) [inline mlflow/store/model_registry/sqlalchemy_store.py:646]:
```suggestion
            src_model_name = parsed_model_uri.name
            src_model_version = parsed_model_uri.version
            try:
                storage_location = self.get_model_version_download_uri(
                    parsed_model_uri.name, parsed_model_uri.version
                )
```

**Review by harupy — APPROVED** [review]

File: mlflow/store/model_registry/abstract_store.py [hunk 8]
```diff
@@ -335,13 +334,17 @@ def copy_model_version(self, src_mv, dst_name):
         :return: Single :py:class:`mlflow.entities.model_registry.ModelVersion` object representing
                  the cloned model version.
         """
-        pass
+        raise MlflowException(
+            "Method 'copy_model_version' has not yet been implemented for the current model "
+            "registry backend. To request support for implementing this method with this backend, "
+            "please submit an issue on GitHub."
+        )
```
Comment (jerrylian-db) [inline mlflow/store/model_registry/abstract_store.py:341]: @harupy before I land this PR, does this error message look good? I discussed with @dbczumar to nudge users to file issues if they would like guidance or support implementing this method for their model registry backend.
